File: cli/skills-catalog.js
const fs = require("fs")
const os = require("os")
const path = require("path")
const { spawnSync } = require("child_process")

function dcliDir() {
  return process.env.SUPERCLI_HOME || path.join(os.homedir(), ".dcli")
}

function supercliDir() {
  return process.env.SUPERCLI_HOME || path.join(os.homedir(), ".supercli")
}

function providersFile() {
  const supercliPath = path.join(supercliDir(), "skills-providers.json")
  const dcliPath = path.join(dcliDir(), "skills-providers.json")
  return fs.existsSync(supercliPath) ? supercliPath : dcliPath
}

function indexFile() {
  const supercliPath = path.join(supercliDir(), "skills-index.json")
  const dcliPath = path.join(dcliDir(), "skills-index.json")
  return fs.existsSync(supercliPath) ? supercliPath : dcliPath
}

function ensureDir() {
  const dir = dcliDir()
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true })
  const supercliDirPath = supercliDir()
  if (!fs.existsSync(supercliDirPath)) fs.mkdirSync(supercliDirPath, { recursive: true })
}

function defaultProviders() {
  const home = os.homedir()
  const cwd = process.cwd()
  return [
    { name: "opencode", type: "local_fs", roots: [path.join(home, ".config", "opencode", "skills")], enabled: true },
    { name: "codex", type: "local_fs", roots: [path.join(home, ".config", "codex", "skills")], enabled: true },
    { name: "windsurf", type: "local_fs", roots: [path.join(home, ".config", "windsurf", "skills")], enabled: true },
    { name: "cursor", type: "local_fs", roots: [path.join(home, ".config", "cursor", "skills")], enabled: true },
    { name: "repo", type: "repo_fs", roots: [path.join(cwd, ".agents", "skills"), path.join(cwd, "docs", "skills")], enabled: true }
  ]
}

function discoverInstalledPluginProviders() {
  const pluginsLockPath = path.join(supercliDir(), "plugins", "plugins.lock.json")
  if (!fs.existsSync(pluginsLockPath)) return []
  
  try {
    const lock = JSON.parse(fs.readFileSync(pluginsLockPath, "utf-8"))
    if (!lock || !lock.installed) return []
    
    const providers = []
    const addedNames = new Set()
    
    for (const [name, plugin] of Object.entries(lock.installed)) {
      if (addedNames.has(name)) continue
      
      const resolvedFrom = plugin.resolved_from || {}
      let pluginDir = null
      
      if (resolvedFrom.type === "bundled") {
        const bundledDir = path.join(__dirname, "..", "plugins")
        pluginDir = path.join(bundledDir, name)
      } else if (resolvedFrom.manifest_path) {
        pluginDir = path.dirname(resolvedFrom.manifest_path)
      }
      
      if (pluginDir && fs.existsSync(pluginDir)) {
        const skillsDirs = [
          path.join(pluginDir, "skills"),
          path.join(pluginDir, ".agents", "skills")
        ]
        const hasSkills = skillsDirs.some(dir => fs.existsSync(dir))
        
        if (hasSkills) {
          providers.push({
            name: name,
            type: "plugin_fs",
            enabled: true,
            plugin_dir: pluginDir
          })
          addedNames.add(name)
        }
      }
    }
    
    return providers
  } catch {
    return []
  }
}

function readJson(file, fallback) {
  try {
    if (!fs.existsSync(file)) return fallback
    return JSON.parse(fs.readFileSync(file, "utf-8"))
  } catch {
    return fallback
  }
}

function writeJson(file, value) {
  ensureDir()
  fs.writeFileSync(file, JSON.stringify(value, null, 2))
  return value
}

function listProviders() {
  const manualProviders = readJson(providersFile(), defaultProviders())
  const manual = Array.isArray(manualProviders) ? manualProviders : defaultProviders()
  const discoveredPlugins = discoverInstalledPluginProviders()
  
  const merged = new Map()
  
  for (const p of manual) {
    merged.set(p.name, p)
  }
  
  for (const p of discoveredPlugins) {
    if (!merged.has(p.name)) {
      merged.set(p.name, p)
    }
  }
  
  return Array.from(merged.values())
}

function setProviders(providers) {
  return writeJson(providersFile(), providers)
}

function addProvider(entry) {
  const providers = listProviders()
  const idx = providers.findIndex(p => p.name === entry.name)
  if (idx >= 0) providers[idx] = entry
  else providers.push(entry)
  setProviders(providers)
  return entry
}

function removeProvider(name) {
  const providers = listProviders()
  const next = providers.filter(p => p.name !== name)
  const removed = next.length !== providers.length
  setProviders(next)
  return removed
}

function getProvider(name) {
  return listProviders().find(p => p.name === name) || null
}

function readIndex() {
  const empty = { version: 1, updated_at: null, skills: [] }
  const idx = readJson(indexFile(), empty)
  if (!idx || typeof idx !== "object") return empty
  if (!Array.isArray(idx.skills)) idx.skills = []
  return idx
}

function walkDir(dir) {
  const files = []
  if (!fs.existsSync(dir)) return files
  const entries = fs.readdirSync(dir, { withFileTypes: true })
  for (const entry of entries) {
    const full = path.join(dir, entry.name)
    if (entry.isDirectory()) {
      files.push(...walkDir(full))
      continue
    }
    if (entry.isFile() && entry.name === "SKILL.md") files.push(full)
  }
  return files
}

function parseFrontmatter(markdown) {
  if (!markdown.startsWith("---\n")) return { frontmatter: {}, body: markdown }
  const end = markdown.indexOf("\n---\n", 4)
  if (end < 0) return { frontmatter: {}, body: markdown }
  const raw = markdown.slice(4, end)
  const body = markdown.slice(end + 5)
  const frontmatter = {}
  for (const line of raw.split("\n")) {
    const idx = line.indexOf(":")
    if (idx < 0) continue
    const key = line.slice(0, idx).trim()
    const value = line.slice(idx + 1).trim().replace(/^"|"$/g, "")
    if (key) frontmatter[key] = value
  }
  return { frontmatter, body }
}

function baseSkillId(filePath, frontmatter, root) {
  if (frontmatter.skill_name) return frontmatter.skill_name
  const rel = path.relative(root, filePath)
  return rel.replace(/\\/g, "/").replace(/\/SKILL\.md$/, "").replace(/\//g, ".")
}

function syncCatalog() {
  const providers = listProviders()
  const skills = []

  for (const provider of providers) {
    if (!provider.enabled) continue

    if (provider.type === "remote_static") {
      const entries = Array.isArray(provider.entries) ? provider.entries : []
      for (const entry of entries) {
        if (!entry || !entry.id || !entry.source_url) continue
        skills.push({
          id: provider.name + ":" + entry.id,
          provider: provider.name,
          name: entry.name || entry.id,
          description: entry.description || "",
          source_url: entry.source_url,
          tags: Array.isArray(entry.tags) ? entry.tags : [],
          updated_at: new Date().toISOString()
        })
      }
      continue
    }

    if (provider.type === "plugin_fs") {
      const pluginDir = provider.plugin_dir
      if (!pluginDir || !fs.existsSync(pluginDir)) continue
      const skillDirs = [
        path.join(pluginDir, "skills"),
        path.join(pluginDir, ".agents", "skills")
      ]
      for (const skillsDir of skillDirs) {
        if (!fs.existsSync(skillsDir)) continue
        const files = walkDir(skillsDir)
        for (const filePath of files) {
          const markdown = fs.readFileSync(filePath, "utf-8")
          const { frontmatter, body } = parseFrontmatter(markdown)
          const baseId = baseSkillId(filePath, frontmatter, skillsDir)
          const heading = body.split("\n").find(l => l.startsWith("# "))
          const name = frontmatter.skill_name || (heading ? heading.slice(2).trim() : baseId)
          const description = frontmatter.description || ""
          const tags = typeof frontmatter.tags === "string" ? frontmatter.tags.split(",").map(t => t.trim()).filter(Boolean) : []
          skills.push({
            id: provider.name + ":" + baseId,
            provider: provider.name,
            name,
            description,
            source_path: filePath,
            tags,
            updated_at: new Date().toISOString()
          })
        }
      }
      continue
    }

    const roots = Array.isArray(provider.roots) ? provider.roots : []
    for (const root of roots) {
      const files = walkDir(root)
      for (const filePath of files) {
        const markdown = fs.readFileSync(filePath, "utf-8")
        const { frontmatter, body } = parseFrontmatter(markdown)
        const baseId = baseSkillId(filePath, frontmatter, root)
        const heading = body.split("\n").find(l => l.startsWith("# "))
        const name = frontmatter.skill_name || (heading ? heading.slice(2).trim() : baseId)
        const description = frontmatter.description || ""
        const tags = typeof frontmatter.tags === "string" ? frontmatter.tags.split(",").map(t => t.trim()).filter(Boolean) : []
        skills.push({
          id: provider.name + ":" + baseId,
          provider: provider.name,
          name,
          description,
          source_path: filePath,
          tags,
          updated_at: new Date().toISOString()
        })
      }
    }
  }

  skills.sort((a, b) => a.id.localeCompare(b.id))
  const next = {
    version: 1,
    updated_at: new Date().toISOString(),
    providers: providers.map(p => p.name),
    skills
  }
  writeJson(indexFile(), next)
  return next
}

function listCatalogSkills(options = {}) {
  const idx = readIndex()
  let items = idx.skills || []
  if (options.provider) items = items.filter(s => s.provider === options.provider)
  return items.map(s => ({ id: s.id, name: s.name, description: s.description, provider: s.provider }))
}

function searchCatalog(query, options = {}) {
  const q = String(query || "").trim().toLowerCase()
  if (!q) return []
  const items = listCatalogSkills(options)
  return items.filter(s =>
    s.id.toLowerCase().includes(q) ||
    s.name.toLowerCase().includes(q) ||
    String(s.description || "").toLowerCase().includes(q)
  )
}

function getCatalogSkill(providerId) {
  const idx = readIndex()
  const hit = (idx.skills || []).find(s => s.id === providerId)
  if (!hit) return null
  let markdown = null
  if (hit.source_path) {
    if (!fs.existsSync(hit.source_path)) return null
    markdown = fs.readFileSync(hit.source_path, "utf-8")
  } else if (hit.source_url) {
    const res = spawnSync("curl", ["-fsSL", hit.source_url], { encoding: "utf-8", timeout: 15000 })
    if (res.error || res.status !== 0) return null
    markdown = (res.stdout || "").trim()
  }
  if (!markdown) return null
  return { ...hit, markdown }
}



/**
 * Get catalog information with provider breakdown
 */
function getCatalogInfo() {
  const idx = readIndex()
  const providers = listProviders()
  
  const providerStats = providers.map(provider => {
    if (!provider.enabled) {
      return { name: provider.name, type: provider.type, enabled: false, status: 'disabled' }
    }
    
    let skillsCount = 0
    let status = 'active'
    let details = {}
    
    if (provider.type === 'plugin_fs') {
      const pluginDir = provider.plugin_dir
      if (!pluginDir || !fs.existsSync(pluginDir)) {
        status = 'missing'
        details.error = 'Plugin directory not found'
      } else {
        const skillsDirs = [
          path.join(pluginDir, 'skills'),
          path.join(pluginDir, '.agents', 'skills')
        ]
        for (const skillsDir of skillsDirs) {
          if (fs.existsSync(skillsDir)) {
            const files = walkDir(skillsDir)
            skillsCount += files.length
          }
        }
        if (skillsCount > 0) {
          details.plugin_dir = pluginDir
          details.skills_dirs = skillsDirs
        } else {
          status = 'no_skills_dir'
          details.skills_dirs = skillsDirs
        }
      }
    } else if (provider.type === 'local_fs' || provider.type === 'repo_fs') {
      const roots = Array.isArray(provider.roots) ? provider.roots : []
      for (const root of roots) {
        if (fs.existsSync(root)) {
          const files = walkDir(root)
          skillsCount += files.length
        }
      }
      details.roots = roots
    } else if (provider.type === 'remote_static') {
      skillsCount = Array.isArray(provider.entries) ? provider.entries.length : 0
      details.entries_count = skillsCount
      details.source_repo = provider.source_repo
    }
    
    return {
      name: provider.name,
      type: provider.type,
      enabled: provider.enabled,
      status,
      skills_count: skillsCount,
      ...details
    }
  })
  
  return {
    index: {
      version: idx.version,
      updated_at: idx.updated_at,
      total_skills: idx.skills.length
    },
    providers: providerStats,
    recent_skills: idx.skills.slice(-5).map(s => ({
      id: s.id,
      name: s.name,
      provider: s.provider,
      added_at: s.updated_at
    }))
  }
}

/**
 * Describe all provider types with examples
 */
function describeProviderTypes() {
  return {
    provider_types: [
      {
        name: 'local_fs',
        description: 'Scans local directories for SKILL.md files',
        example: {
          name: 'my-skills',
          type: 'local_fs',
          roots: ['~/.config/opencode/skills'],
          enabled: true
        }
      },
      {
        name: 'repo_fs',
        description: 'Scans repository directories for SKILL.md files',
        example: {
          name: 'repo',
          type: 'repo_fs',
          roots: ['.agents/skills', 'docs/skills'],
          enabled: true
        }
      },
      {
        name: 'remote_static',
        description: 'Pre-indexed skills from remote repositories (GitHub)',
        example: {
          name: 'visual-explainer',
          type: 'remote_static',
          source_repo: 'https://github.com/user/repo',
          entries: [{ id: 'skill-1', source_url: 'https://raw.githubusercontent.com/...' }]
        }
      },
      {
        name: 'plugin_fs',
        description: 'Auto-discovers SKILL.md files from installed plugin skills/ directory',
        example: {
          name: 'superbackend',
          type: 'plugin_fs',
          plugin_dir: '/path/to/plugins/superbackend'
        }
      }
    ]
  }
}


module.exports = {
  listProviders,
  addProvider,
  removeProvider,
  getProvider,
  readIndex,
  syncCatalog,
  listCatalogSkills,
  searchCatalog,
  getCatalogSkill,
  getCatalogInfo,
  describeProviderTypes
}

File: supercli-zig-cli/src/skills_catalog.zig
// skills_catalog.zig — Skill providers, index management, and SKILL.md scanning.
// Mirrors cli/skills-catalog.js for local_fs/repo_fs/plugin_fs providers.
const std = @import("std");

pub const SkillEntry = struct {
    id: []const u8,
    provider: []const u8,
    name: []const u8,
    description: []const u8 = "",
    source_path: []const u8 = "",
    tags: [][]const u8 = &.{},
};

pub const Provider = struct {
    name: []const u8,
    type: []const u8,
    roots: [][]const u8 = &.{},
    enabled: bool = true,
    plugin_dir: []const u8 = "",
};

fn supercliDir(gpa: std.mem.Allocator, home: []const u8) ![]const u8 {
    return std.fmt.allocPrint(gpa, "{s}/.supercli", .{home});
}

fn providersFile(gpa: std.mem.Allocator, home: []const u8) ![]const u8 {
    const dir = try supercliDir(gpa, home);
    return std.fmt.allocPrint(gpa, "{s}/skills-providers.json", .{dir});
}

fn indexFile(gpa: std.mem.Allocator, home: []const u8) ![]const u8 {
    const dir = try supercliDir(gpa, home);
    return std.fmt.allocPrint(gpa, "{s}/skills-index.json", .{dir});
}

/// Default providers (matches Node.js defaults).
pub fn defaultProviders(gpa: std.mem.Allocator, home: []const u8) ![]Provider {
    var list: std.ArrayList(Provider) = .empty;

    const configs = [_]struct { name: []const u8, sub: []const u8 }{
        .{ .name = "opencode", .sub = ".config/opencode/skills" },
        .{ .name = "codex", .sub = ".config/codex/skills" },
        .{ .name = "windsurf", .sub = ".config/windsurf/skills" },
        .{ .name = "cursor", .sub = ".config/cursor/skills" },
    };
    for (configs) |c| {
        const root = try std.fmt.allocPrint(gpa, "{s}/{s}", .{ home, c.sub });
        const roots = try gpa.alloc([]const u8, 1);
        roots[0] = root;
        try list.append(gpa, .{ .name = c.name, .type = "local_fs", .roots = roots });
    }

    // repo provider (cwd-based, use home as fallback)
    const repo_root = try std.fmt.allocPrint(gpa, "{s}/.agents/skills", .{home});
    const repo_roots = try gpa.alloc([]const u8, 1);
    repo_roots[0] = repo_root;
    try list.append(gpa, .{ .name = "repo", .type = "repo_fs", .roots = repo_roots });

    // supercli provider (bundled skills)
    const sc_root = try std.fmt.allocPrint(gpa, "{s}/ai/supercli/.agents/skills", .{home});
    const sc_roots = try gpa.alloc([]const u8, 1);
    sc_roots[0] = sc_root;
    try list.append(gpa, .{ .name = "supercli", .type = "local_fs", .roots = sc_roots });

    return list.toOwnedSlice(gpa);
}

/// Read providers from file, merged with defaults.
pub fn listProviders(gpa: std.mem.Allocator, io: std.Io, home: []const u8) ![]Provider {
    const defaults = try defaultProviders(gpa, home);

    const pf = try providersFile(gpa, home);
    const content = std.Io.Dir.cwd().readFileAlloc(io, pf, gpa, .unlimited) catch return defaults;

    const parsed = std.json.parseFromSliceLeaky(std.json.Value, gpa, content, .{}) catch return defaults;
    const arr = switch (parsed) {
        .array => parsed.array.items,
        else => return defaults,
    };

    var merged: std.ArrayList(Provider) = .empty;
    for (defaults) |d| try merged.append(gpa, d);

    for (arr) |item| switch (item) {
        .object => |obj| {
            const name = if (obj.get("name")) |v| switch (v) {
                .string => |s| try gpa.dupe(u8, s),
                else => continue,
            } else continue;

            const ptype = if (obj.get("type")) |v| switch (v) {
                .string => |s| try gpa.dupe(u8, s),
                else => "local_fs",
            } else "local_fs";
            const enabled = if (obj.get("enabled")) |v| switch (v) {
                .bool => |b| b,
                else => true,
            } else true;

            // Parse roots array
            var roots: [][]const u8 = &.{};
            if (obj.get("roots")) |rv| switch (rv) {
                .array => |ra| {
                    roots = try gpa.alloc([]const u8, ra.items.len);
                    for (ra.items, 0..) |ri, ri_i| {
                        roots[ri_i] = switch (ri) {
                            .string => |s| try gpa.dupe(u8, s),
                            else => "",
                        };
                    }
                },
                else => {},
            };

            // Parse plugin_dir
            var plugin_dir: []const u8 = "";
            if (obj.get("plugin_dir")) |v| switch (v) {
                .string => |s| plugin_dir = try gpa.dupe(u8, s),
                else => {},
            };

            // Override if exists
            var found = false;
            for (merged.items, 0..) |m, i| {
                if (std.mem.eql(u8, m.name, name)) {
                    found = true;
                    merged.items[i] = .{ .name = name, .type = ptype, .roots = roots, .enabled = enabled, .plugin_dir = plugin_dir };
                    break;
                }
            }
            if (!found) {
                try merged.append(gpa, .{ .name = name, .type = ptype, .roots = roots, .enabled = enabled, .plugin_dir = plugin_dir });
            }
        },
        else => {},
    };

    return merged.toOwnedSlice(gpa);
}

/// Walk a directory recursively, collecting SKILL.md file paths.
fn walkDir(gpa: std.mem.Allocator, io: std.Io, dir_path: []const u8, list: *std.ArrayList([]const u8)) !void {
    var dir = std.Io.Dir.cwd().openDir(io, dir_path, .{ .iterate = true }) catch return;
    defer dir.close(io);

    var iter = dir.iterate();
    while (try iter.next(io)) |entry| {
        if (entry.kind == .directory) {
            const sub = try std.fmt.allocPrint(gpa, "{s}/{s}", .{ dir_path, entry.name });
            try walkDir(gpa, io, sub, list);
        } else if (entry.kind == .file and std.mem.eql(u8, entry.name, "SKILL.md")) {
            const full = try std.fmt.allocPrint(gpa, "{s}/{s}", .{ dir_path, entry.name });
            try list.append(gpa, full);
        }
    }
}

/// Parse YAML frontmatter from SKILL.md markdown.
fn parseFrontmatter(gpa: std.mem.Allocator, markdown: []const u8) !struct { name: []const u8, description: []const u8 } {
    var name: []const u8 = "";
    var description: []const u8 = "";

    // Check for --- prefix
    if (!std.mem.startsWith(u8, markdown, "---\n")) return .{ .name = name, .description = description };

    // Find closing ---
    const end_marker = "\n---\n";
    const end_pos = std.mem.indexOfPos(u8, markdown, 4, end_marker) orelse return .{ .name = name, .description = description };
    const raw = markdown[4..end_pos];

    var line_iter = std.mem.tokenizeAny(u8, raw, "\n");
    while (line_iter.next()) |line| {
        const colon = std.mem.indexOf(u8, line, ":") orelse continue;
        const key = std.mem.trim(u8, line[0..colon], " ");
        const value_raw = std.mem.trim(u8, line[colon + 1 ..], " ");
        // Strip surrounding quotes
        var value = value_raw;
        if (value.len >= 2 and value[0] == '"' and value[value.len - 1] == '"') {
            value = value[1 .. value.len - 1];
        }
        if (std.mem.eql(u8, key, "skill_name") or std.mem.eql(u8, key, "name")) {
            name = try gpa.dupe(u8, value);
        } else if (std.mem.eql(u8, key, "description")) {
            description = try gpa.dupe(u8, value);
        }
    }

    // Fallback: find first # heading in body
    if (name.len == 0) {
        const body = markdown[end_pos + end_marker.len ..];
        var body_iter = std.mem.tokenizeAny(u8, body, "\n");
        while (body_iter.next()) |bline| {
            if (std.mem.startsWith(u8, bline, "# ")) {
                name = try gpa.dupe(u8, std.mem.trim(u8, bline[2..], " "));
                break;
            }
        }
    }

    return .{ .name = name, .description = description };
}

/// Derive skill ID from file path relative to root.
fn deriveSkillId(gpa: std.mem.Allocator, file_path: []const u8, root: []const u8) ![]const u8 {
    // Remove root prefix and /SKILL.md suffix
    var rel = file_path;
    if (std.mem.startsWith(u8, rel, root)) {
        rel = rel[root.len..];
        if (rel.len > 0 and rel[0] == '/') rel = rel[1..];
    }
    // Remove trailing /SKILL.md
    if (std.mem.endsWith(u8, rel, "/SKILL.md")) {
        rel = rel[0 .. rel.len - "/SKILL.md".len];
    } else if (std.mem.endsWith(u8, rel, "SKILL.md")) {
        rel = rel[0 .. rel.len - "SKILL.md".len];
    }
    // Replace / with .
    const result = try gpa.alloc(u8, rel.len);
    for (rel, 0..) |c, i| {
        result[i] = if (c == '.') '-' else c;
    }
    // Now replace / with .
    for (result, 0..) |c, i| {
        if (c == '/') result[i] = '.';
    }
    return result;
}

/// Sync catalog: scan all enabled providers for SKILL.md files, build index.
pub fn syncCatalog(gpa: std.mem.Allocator, io: std.Io, home: []const u8) !struct { skills: []SkillEntry, providers: [][]const u8 } {
    const providers = try listProviders(gpa, io, home);

    var skills: std.ArrayList(SkillEntry) = .empty;
    var provider_names: std.ArrayList([]const u8) = .empty;

    for (providers) |provider| {
        if (!provider.enabled) continue;
        try provider_names.append(gpa, provider.name);

        // For local_fs and repo_fs: scan roots
        for (provider.roots) |root| {
            var files: std.ArrayList([]const u8) = .empty;
            try walkDir(gpa, io, root, &files);

            for (files.items) |file_path| {
                const content = std.Io.Dir.cwd().readFileAlloc(io, file_path, gpa, .unlimited) catch continue;
                const fm = try parseFrontmatter(gpa, content);

                const base_id = try deriveSkillId(gpa, file_path, root);
                const id = try std.fmt.allocPrint(gpa, "{s}:{s}", .{ provider.name, base_id });

                try skills.append(gpa, .{
                    .id = id,
                    .provider = provider.name,
                    .name = if (fm.name.len > 0) fm.name else base_id,
                    .description = fm.description,
                    .source_path = file_path,
                });
            }
        }
    }

    // Write index file
    const idx_path = try indexFile(gpa, home);
    var out: std.Io.Writer.Allocating = .init(gpa);
    defer out.deinit();
    var jw: std.json.Stringify = .{ .writer = &out.writer, .options = .{ .whitespace = .indent_2 } };
    try jw.beginObject();
    try jw.objectField("version");
    try jw.write(1);
    try jw.objectField("updated_at");
    try jw.write(null);
    try jw.objectField("providers");
    try jw.beginArray();
    for (provider_names.items) |pn| try jw.write(pn);
    try jw.endArray();
    try jw.objectField("skills");
    try jw.beginArray();
    for (skills.items) |s| {
        try jw.beginObject();
        try jw.objectField("id");
        try jw.write(s.id);
        try jw.objectField("provider");
        try jw.write(s.provider);
        try jw.objectField("name");
        try jw.write(s.name);
        try jw.objectField("description");
        try jw.write(s.description);
        try jw.objectField("source_path");
        try jw.write(s.source_path);
        try jw.endObject();
    }
    try jw.endArray();
    try jw.endObject();

    try std.Io.Dir.cwd().writeFile(io, .{ .sub_path = idx_path, .data = out.written() });

    return .{ .skills = try skills.toOwnedSlice(gpa), .providers = try provider_names.toOwnedSlice(gpa) };
}

/// Read the skills index file.
pub fn readIndex(gpa: std.mem.Allocator, io: std.Io, home: []const u8) ![]SkillEntry {
    const idx_path = try indexFile(gpa, home);
    const content = std.Io.Dir.cwd().readFileAlloc(io, idx_path, gpa, .unlimited) catch return &.{};

    const parsed = std.json.parseFromSliceLeaky(std.json.Value, gpa, content, .{}) catch return &.{};
    const root = switch (parsed) {
        .object => parsed.object,
        else => return &.{},
    };

    const skills_arr = root.get("skills") orelse return &.{};
    const arr = switch (skills_arr) {
        .array => skills_arr.array.items,
        else => return &.{},
    };

    var list: std.ArrayList(SkillEntry) = .empty;
    for (arr) |item| switch (item) {
        .object => |obj| {
            const id = if (obj.get("id")) |v| switch (v) {
                .string => |s| try gpa.dupe(u8, s),
                else => continue,
            } else continue;
            const provider = if (obj.get("provider")) |v| switch (v) {
                .string => |s| try gpa.dupe(u8, s),
                else => "",
            } else "";
            const name = if (obj.get("name")) |v| switch (v) {
                .string => |s| try gpa.dupe(u8, s),
                else => "",
            } else "";
            const description = if (obj.get("description")) |v| switch (v) {
                .string => |s| try gpa.dupe(u8, s),
                else => "",
            } else "";
            const source_path = if (obj.get("source_path")) |v| switch (v) {
                .string => |s| try gpa.dupe(u8, s),
                else => "",
            } else "";

            try list.append(gpa, .{
                .id = id,
                .provider = provider,
                .name = name,
                .description = description,
                .source_path = source_path,
            });
        },
        else => {},
    };

    return list.toOwnedSlice(gpa);
}

/// Search catalog skills by query substring (case-insensitive).
pub fn searchCatalog(gpa: std.mem.Allocator, io: std.Io, home: []const u8, query: []const u8) ![]SkillEntry {
    const all = try readIndex(gpa, io, home);
    if (query.len == 0) return all;

    const q_lower = try std.ascii.allocLowerString(gpa, query);
    defer gpa.free(q_lower);

    var filtered: std.ArrayList(SkillEntry) = .empty;
    for (all) |s| {
        const id_lower = try std.ascii.allocLowerString(gpa, s.id);
        defer gpa.free(id_lower);
        const name_lower = try std.ascii.allocLowerString(gpa, s.name);
        defer gpa.free(name_lower);
        const desc_lower = try std.ascii.allocLowerString(gpa, s.description);
        defer gpa.free(desc_lower);

        if (std.mem.indexOf(u8, id_lower, q_lower) != null or
            std.mem.indexOf(u8, name_lower, q_lower) != null or
            std.mem.indexOf(u8, desc_lower, q_lower) != null)
        {
            try filtered.append(gpa, s);
        }
    }
    return filtered.toOwnedSlice(gpa);
}

/// Get a specific skill by provider:id format.
pub fn getCatalogSkill(gpa: std.mem.Allocator, io: std.Io, home: []const u8, skill_id: []const u8) !?struct { entry: SkillEntry, markdown: []const u8 } {
    const all = try readIndex(gpa, io, home);
    for (all) |s| {
        if (std.mem.eql(u8, s.id, skill_id)) {
            if (s.source_path.len > 0) {
                const content = std.Io.Dir.cwd().readFileAlloc(io, s.source_path, gpa, .unlimited) catch return null;
                return .{ .entry = s, .markdown = content };
            }
            return null;
        }
    }
    return null;
}
